lib/vrl/compiler/src/expression/function_call.rs

```diff
@@ -364,12 +370,25 @@ impl<'a> Builder<'a> {
         local: &mut LocalEnv,
         external: &mut ExternalEnv,
         closure_block: Option<Node<Block>>,
+        mut local_snapshot: LocalEnv,
     ) -> Result<FunctionCall, Error> {
         let mut closure_fallible = false;
 
         // Check if we have a closure we need to compile.
         if let Some((variables, input)) = self.closure.clone() {
             let block = closure_block.expect("closure must contain block");
+
+            // At this point, we've compiled the block, so we can remove the
+            // closure variables from the compiler's local environment.
+            variables
+                .iter()
+                .for_each(|ident| match local_snapshot.remove_variable(ident) {
+                    Some(details) => local.insert_variable(ident.clone(), details),
+                    None => {
+                        local.remove_variable(ident);
+                    }
+                });
```

Comment:
Correct. I could match against `local_snapshot.variable(ident)`, but then I'd have to clone `details`, and since we don't need `local_snapshot` any more, there's no harm in taking the details by-value.

I need the `remove_variable` API on line 388 regardless, so it wouldn't get rid of that method implementation if we went that route.